### PR TITLE
fix: validate payout ledger requests

### PR DESCRIPTION
**Pull request by cerredz**
## Summary
- require payout ledger write endpoints to receive JSON objects instead of scalars/arrays/null
- reject non-numeric and non-finite `amount_rtc` values before inserting ledger rows
- cover invalid request bodies plus the existing create/status happy path

## Verification
- `python -m pytest tests\test_payout_ledger_request_validation.py -q`
- `python -m pytest tests\test_payout_ledger_migration.py tests\test_payout_ledger_request_validation.py -q`
- `python -m py_compile tests\test_payout_ledger_request_validation.py payout_ledger.py node\utxo_db.py`
- `git diff --check -- tests\test_payout_ledger_request_validation.py payout_ledger.py node\utxo_db.py`
- `python -m pytest tests\security_audit\test_security_findings_2867.py::test_mempool_add_manage_tx_undefined -q`

### CODE REVIEW

**Review by douglance — CHANGES_REQUESTED** [review]
Thanks for tightening the payout ledger write routes. The object-shape checks and non-finite handling are moving in the right direction, but I found one validation bypass that should be closed before merge.

`_parse_amount_rtc()` currently accepts JSON booleans because Python treats `bool` as a numeric subtype for `float()` conversion. A request body with `"amount_rtc": true` reaches `ledger_create()` and stores `1.0` RTC; `false` stores `0.0`. Since the PR goal is to reject non-numeric `amount_rtc` values, booleans should be rejected before coercion.

Reproduction against this branch:
- `POST /api/ledger` with `{"bounty_id":"bounty-1","contributor":"alice","amount_rtc":true}` -> `201 {'id': '...', 'status': 'queued'}`

Validation I ran:
- `uv run --no-project --with pytest --with flask python -m pytest tests/test_payout_ledger_request_validation.py -q` -> 6 passed
- one-off Flask test-client reproduction above -> boolean amount accepted as a queued payout
- `python3 -m py_compile tests/test_payout_ledger_request_validation.py payout_ledger.py node/utxo_db.py` -> passed
- `git diff --check origin/main...HEAD -- tests/test_payout_ledger_request_validation.py payout_ledger.py node/utxo_db.py` -> passed

A minimal fix would be to explicitly reject `isinstance(value, bool)` before calling `float(value)`, and add boolean cases to the invalid amount regression coverage.

File: payout_ledger.py [hunk 1]
```diff
@@ -12,12 +12,30 @@
...
+        return None, (jsonify({"error": "JSON object required"}), 400)
+    return data, None
+
+
+def _parse_amount_rtc(value):
+    try:
+        amount = float(value)
+    except (TypeError, ValueError):
+        return None, (jsonify({"error": "amount_rtc must be a finite number"}), 400)
+    if not math.isfinite(amount):
```
Comment (BrianYeh) [inline payout_ledger.py]: This only rejects non-finite values. A finite zero or negative value still falls through to `return amount, None`, so `/api/ledger` can create queued payout rows with `amount_rtc` like `0` or `-5`. The repo payout preflight docs call out non-positive `amount_rtc` as invalid, so this should also check `amount <= 0` before accepting the request.

**Review by BrianYeh — CHANGES_REQUESTED** [review]
I found one blocking validation gap in the payout ledger request hardening. The patch now rejects non-object JSON and non-finite `amount_rtc`, but it still accepts finite zero or negative payout amounts. That conflicts with the repository's payout preflight expectation for non-positive amounts and lets API callers create queued payout records with values like `-5` RTC.

Checks performed:
- `py -m py_compile payout_ledger.py tests/test_payout_ledger_request_validation.py`
- `git diff --check 09cd06f9e28d4027bcb3992d1b43fe246fe6eadf...HEAD -- payout_ledger.py tests/test_payout_ledger_request_validation.py`
- Tried `py -m pytest tests/test_payout_ledger_request_validation.py tests/test_payout_ledger_migration.py`; local execution is blocked because pytest is not installed in this environment.
- Tried a Flask client reproduction for `amount_rtc=-5`; local execution is blocked by missing Flask, but the static path is direct: `_parse_amount_rtc()` returns any finite float unchanged and `api_ledger_create()` passes it to `ledger_create()`.

Suggested fix: reject `amount <= 0` in `_parse_amount_rtc` and add a regression test for `0`, negative values, and maybe `-0.0`.

**Comment by cerredz** [comment]
Addressed both review findings in `202ebd6`:

- reject JSON booleans before `float()` coercion so `true`/`false` cannot become `1.0`/`0.0`
- reject non-positive finite values (`0`, negative numbers, `-0.0`) before ledger insertion
- extended `tests/test_payout_ledger_request_validation.py` to cover booleans, zero, negative, and non-finite/invalid values

Local verification after the update:
- `python -m pytest tests\test_payout_ledger_request_validation.py tests\test_payout_ledger_migration.py -q` -> 13 passed
- `python -m py_compile tests\test_payout_ledger_request_validation.py payout_ledger.py node\utxo_db.py` -> passed
- `git diff --check -- tests\test_payout_ledger_request_validation.py payout_ledger.py node\utxo_db.py` -> passed
- `python -m pytest tests\security_audit\test_security_findings_2867.py::test_mempool_add_manage_tx_undefined -q` -> passed with the existing return-value warning

**Review by BrianYeh — APPROVED** [review]
Re-reviewed after the new commit `202ebd6` (`fix: reject invalid payout ledger amounts`). The blocker I raised in my previous review is resolved: `_parse_amount_rtc()` now rejects booleans, non-finite values, zero, negative values, and `-0.0`, and the regression test parametrization covers those cases.

Checks performed on the updated head:
- `py -m py_compile payout_ledger.py tests/test_payout_ledger_request_validation.py`
- `git diff --check 09cd06f9e28d4027bcb3992d1b43fe246fe6eadf...HEAD -- payout_ledger.py tests/test_payout_ledger_request_validation.py`
- Tried `py -m pytest tests/test_payout_ledger_request_validation.py tests/test_payout_ledger_migration.py`; local execution is still blocked because pytest is not installed in this environment.

No remaining blocker from my prior finding.

**Review by douglance — APPROVED** [review]
Follow-up review after the latest validation update: the blocker I raised is fixed.

What I checked:
- `amount_rtc` booleans are now rejected before float coercion.
- non-finite values, zero, negative values, and `-0.0` are rejected.
- valid positive numeric strings still create queued ledger records.
- status-update JSON-object validation remains unchanged.

Local validation:
- `uv run --no-project --with pytest --with flask python -m pytest tests/test_payout_ledger_request_validation.py -q` -> 11 passed
- `python3 -m py_compile tests/test_payout_ledger_request_validation.py payout_ledger.py node/utxo_db.py` -> passed
- `git diff --check origin/main...HEAD -- tests/test_payout_ledger_request_validation.py payout_ledger.py node/utxo_db.py` -> passed

**Review by godd-ctrl — APPROVED** [review]
Reviewed the payout-ledger request validation hardening against the updated PR head.

The current patch correctly rejects non-object JSON bodies on the ledger write endpoints and validates `amount_rtc` before queueing payouts. The follow-up commit closes the two important coercion gaps raised earlier: booleans are rejected before Python can coerce them to `1.0` or `0.0`, and zero/negative values including `-0.0` are rejected before insertion.

I also checked the status-update path. It still accepts the intended object payloads while arrays, scalars, and invalid status values fail with controlled 400 responses instead of reaching ledger mutation logic.

Local validation:
- `uv run --no-project --with pytest --with flask python -m pytest tests\test_payout_ledger_request_validation.py tests\test_payout_ledger_migration.py -q` -> 13 passed
- `uv run --no-project --with pytest --with flask python -m pytest tests\security_audit\test_security_findings_2867.py::test_mempool_add_manage_tx_undefined -q` -> 1 passed with the existing PytestReturnNotNoneWarning
- `python -m py_compile tests\test_payout_ledger_request_validation.py payout_ledger.py node\utxo_db.py` -> passed
- `git diff --check origin/main...HEAD -- tests\test_payout_ledger_request_validation.py payout_ledger.py node\utxo_db.py` -> passed

No blockers found on the updated head.

**Review by jaxint — APPROVED** [review]
LGTM! Nice work. Approved ✅

**Review by jaxint — APPROVED** [review]
LGTM! Thanks for contributing. Approved.

**Review by saim256 — CHANGES_REQUESTED** [review]
The validation logic looks functionally sound on the current head, including the earlier amount-validation blockers, but this PR still fails the repository SPDX gate.

Local checks run:
- `python -m pytest tests\test_payout_ledger_request_validation.py -q` -> 11 passed
- `python -m pytest tests\test_payout_ledger_migration.py tests\test_payout_ledger_request_validation.py -q` -> 13 passed
- `python -m py_compile tests\test_payout_ledger_request_validation.py payout_ledger.py` -> passed
- `git diff --check origin/main...HEAD -- tests/test_payout_ledger_request_validation.py payout_ledger.py` -> passed
- `python tools\bcos_spdx_check.py --base-ref origin/main` -> failed on `tests/test_payout_ledger_request_validation.py`

The request-shape and amount handling now reject non-object JSON, booleans, non-finite values, zero, and negative amounts, while valid create/status update still works. Please add the required SPDX header to the new test file so the repo gate passes.

**Comment by cerredz** [comment]
Addressed the requested hygiene blocker on the current head:

- added `# SPDX-License-Identifier: MIT` to `tests/test_payout_ledger_request_validation.py`

Verification after the fix:
- `python -m pytest tests\test_payout_ledger_request_validation.py -q` -> 11 passed
- `python -m pytest tests\test_payout_ledger_migration.py tests\test_payout_ledger_request_validation.py -q` -> 13 passed
- `python -m py_compile tests\test_payout_ledger_request_validation.py payout_ledger.py` -> passed
- `python tools\bcos_spdx_check.py --base-ref origin/main` -> OK
- `git diff --check -- tests\test_payout_ledger_request_validation.py payout_ledger.py` -> passed

**Review by jaxint — APPROVED** [review]
LGTM! Thanks for contributing. Approved.

**Review by jaxint — APPROVED** [review]
LGTM! Thanks for contributing. Approved.

**Review by jaxint — APPROVED** [review]
LGTM! Thanks for contributing. Approved.

**Review by jaxint — APPROVED** [review]
LGTM! Thanks for contributing. Approved.

**Review by loganoe — APPROVED** [review]
Reviewed the payout ledger request validation. The new object-shape guard prevents non-dict JSON from reaching `.get()`, amount parsing now rejects booleans, non-finite values, zero, and negatives before persistence, and the existing create/update paths still work against the migrated table.

Validation run locally:
- `PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 PYTHONPATH=. /tmp/rustchain-flask-venv/bin/python -m pytest -q tests/test_payout_ledger_request_validation.py`
- `PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 PYTHONPATH=. /tmp/rustchain-flask-venv/bin/python -m pytest -q tests/test_payout_ledger_migration.py tests/test_payout_ledger_request_validation.py`
- `/tmp/rustchain-flask-venv/bin/python -m py_compile payout_ledger.py tests/test_payout_ledger_request_validation.py`

**Review by TJCurnutte — APPROVED** [review]
Bounty #73 review — APPROVED

I reviewed this focused RustChain PR and validated the changed files before approving.

Scope inspected:
- PR: #4565 — fix: validate payout ledger requests
- Changed files: payout_ledger.py, tests/test_payout_ledger_request_validation.py
- Diff size: 2 files changed, 118 insertions(+), 3 deletions(-)

Validation evidence:
- `git diff --check origin/main...HEAD` — passed
- `python3 tools/bcos_spdx_check.py --base-ref origin/main` — passed
- Added-line secret-pattern scan — passed
- `python3 -B -m py_compile payout_ledger.py tests/test_payout_ledger_request_validation.py` — passed
- `python3 -B -m pytest -q tests/test_payout_ledger_request_validation.py` — passed

Review reasoning:
- The diff is scoped to `payout_ledger.py, tests/test_payout_ledger_request_validation.py` and matches the PR intent.
- Whitespace/conflict checks are clean, no new unlicensed code files were introduced, and no added secret-like literals were detected.
- The applicable syntax/test gates passed locally, so I did not find a merge-blocking issue in this review.

**Comment by guangningsun** [comment]
## PR Review — Standard Quality ✓

**PR:** #4565 — Fix: validate payout ledger requests

### What I reviewed
- `payout_ledger.py`
- `tests/test_payout_ledger_request_validation.py` (new 88-line test)

### Observations

1. **Adding `import math` suggests validation using `math.isnan()` or `math.isinf()`** — these functions ensure numeric values are valid (not NaN or Infinity), which is important for financial calculations.

2. **New 88-line test validates request shapes** — comprehensive testing of the payout ledger's input validation.

3. **Payout ledger handles miner rewards** — validating all requests prevents bad payouts from being processed.

LGTM.

**Bounty:** #2782
**Disclosure:** I received RTC compensation for this review.

**Review by jaxint — APPROVED** [review]
## PR Review Summary

**LGTM** ✅ — Solid contribution to the RustChain ecosystem.

### Code Quality
- Implementation follows project conventions
- Security considerations adequately addressed  
- Error handling appears robust

### Testing
- Test coverage is appropriate for the changes
- Edge cases covered

**Approved.** Ready to merge. 🚀

**Review by jaxint — APPROVED** [review]
LGTM! Thanks for contributing. Approved.

**Comment by Scottcjn** [comment]
Heads up — codex audit gave this PR a thumbs up (low severity, 10 RTC queued: *Rejects non-object JSON or malformed typed input while preserving auth.*), but the branch now conflicts with `main` after the rest of your cluster merged. GitHub auto-rebase couldn't resolve it (real overlap on the same files).

Could you rebase against the current `main` tip? Once green, this is good to go. If easier, close + reopen a fresh PR off the latest tip — also fine.

Thanks for the cluster work overall — 57 of your 99 PRs merged this round.

**Review by HCIE2054 — APPROVED** [review]
LGTM!

**Review by HCIE2054 — APPROVED** [review]
LGTM! Great contribution to the RustChain ecosystem. Thanks for keeping the codebase clean and well-tested. Approved ✅

**Review by HCIE2054 — APPROVED** [review]
LGTM!

**Review by HCIE2054 — APPROVED** [review]

**Review by 508704820 — COMMENTED** [review]
Validate payout ledger requests. Verify: (1) Payout amount is validated before ledger entry. (2) Request cannot create negative balance. (3) Duplicate payout detection. - Xeophon (security review, payout)

**Review by HCIE2054 — APPROVED** [review]
LGTM!

**Comment by Scottcjn** [comment]
Closing — your branch has irreconcilable merge conflicts (week-old, base has moved). Confirmed as **first-poster credit-eligible** for this fix per Codex forensic audit 2026-05-18 (cerredz: payout ledger request-shape validation (audit Q8 MEDIUM)). The work is recognized; bounty credit stands. 

If you'd like the canonical version of this fix merged from your branch instead of a later submission, rebase and ping — happy to re-evaluate. Otherwise this is closed administratively.

**Review by BossChaos — COMMENTED** [review]
## PR #4565 Review - Ledger API JSON + amount_rtc validation

### Security Analysis - Input Validation

This PR adds JSON object and amount validation to the payout ledger API.

**Key improvements:**
1. **_get_json_object()**: Validates JSON body is dict
2. **_parse_amount_rtc()**: Validates amount is finite positive number, rejects booleans
3. **math.isfinite()**: Prevents NaN/Infinity values
4. **isinstance(value, bool)**: Explicit boolean rejection before float conversion

**Recommendation**: Merge - **important security fix**. Prevents financial amount manipulation.
